src/config.js

```diff
@@ -1,82 +1,179 @@
-"use strict";
-
-var assert      = require('assert');
-var _           = require('lodash');
-var nconf       = require('nconf');
+let _ = require('lodash');
+let yaml = require('js-yaml');
+let fs = require('fs');
+let debug = require('debug')('taskcluster-lib-config');
+let assert = require('assert');
 
 /**
- * Load and return a configuration object (nconf, not global)
- * Access properties using `config.get('object:key')`.
+ * Load configuration from files and environment variables.
  *
  * options:
+ * ```js
  * {
- *   defaults:     {},   // Default configuration values
- *   profile:      {},   // Profile configuration values (overrides defaults)
- *   envs:         [],   // Values to load from env, using `_` as separator
- *   filename:     null  // Filename to load config from at multiple levels
+ *   files: [               // Files to load configuration from
+ *     'config.yml',        // Defaults are relative to process.cwd
+ *     'user-config.yml'
+ *   ]
+ *   profile:  undefined,   // Profile to apply
+ *   env:      process.env  // Environment variables (mapping string to strings)
  * }
+ * ```
+ *
+ * Configuration Format:
+ * ```yaml
+ * Defaults:
+ *   hostname:     localhost
+ *   port:         8080
+ * Profiles:
+ *   production:
+ *     hostname:   example.com
+ *     port:       !env:number PORT
+ * ```
+ *
+ * The following special YAML types can be used to load from environment
+ * variables:
+ * ```
+ * !env        <NAME>  Load string from env var <NAME>
+ * !env:string <NAME>  Load string from env var <NAME>
+ * !env:number <NAME>  Load number from env var <NAME>
+ * !env:flag   <NAME>  Load true if env var <NAME> is defined
+ * !env:bool   <NAME>  Load boolean as /true/i or /false/i from env var <NAME>
+ * !env:json   <NAME>  Load JSON object from env var <NAME>
+ * !env:list   <NAME>  Load list of space separated strings from env var <NAME>
+ * ```
  *
- * Configuration values are searched in the following order:
- *  1. Environment variables
- *  2. Current working director (`./<filename>.conf.json`),
- *  3. Home folder (`~/.<filename>.conf.json`), and
- *  4. System config (`/etc/<filename>.conf.json`)
- *  5. Default value from `profile`
- *  6. Default value from `defaults`
+ * If the environment variable in question isn't defined, the value will be
+ * `undefined`, so it can fall-back to defaults from previous config file.
  */
-var config = function(options) {
-  // Set default options
+let config = (options) => {
```

Comment:
I understand your reasoning for using let to ensure that the function isn't used before declared, but I do personally find find using an arrow function instead of just `function()` kind of confusing here.  Arrow functions are great as parameters to functions, but as a top level exported thing, it just sort of blends in to much for me personally.

Not going to block on this though.